akka-typed/src/main/scala/akka/typed/persistence/scaladsl/PersistentActor.scala

```diff
@@ -20,20 +20,29 @@ object PersistentActor {
       recoveryCompleted = (state, _) ⇒ state)
 
   sealed abstract class PersistentEffect[+Event, State]() {
-    def andThen(callback: State ⇒ Unit): PersistentEffect[Event, State]
-  }
 
-  final case class PersistNothing[Event, State](callbacks: List[State ⇒ Unit] = Nil) extends PersistentEffect[Event, State] {
-    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+    def andThen(sideEffect: ChainableEffect[_, State]): PersistentEffect[Event, State] =
+      CompositeEffect(this, List(sideEffect))
+
+    /** Convenience method to register a side effect with just a callback function */
+    def andThen(callback: State ⇒ Unit): PersistentEffect[Event, State] = andThen(SideEffect(callback))
   }
 
-  case class Persist[Event, State](event: Event, callbacks: List[State ⇒ Unit] = Nil) extends PersistentEffect[Event, State] {
-    def andThen(callback: State ⇒ Unit) = copy(callbacks = callback :: callbacks)
+  case class CompositeEffect[Event, State](mainEffect: PersistentEffect[Event, State], sideEffects: List[ChainableEffect[_, State]]) extends PersistentEffect[Event, State] {
+    override def andThen(sideEffect: ChainableEffect[_, State]): PersistentEffect[Event, State] =
+      copy(sideEffects = sideEffects :+ sideEffect)
```

Comment:
could we prepend (since we use List, and that is good since we expect few elements), and reverse before executing?

Reply:
could something like this be simpler?
```
sealed abstract class PersistentEffect[Event, State]() {
  def andThen(effect: PersistentEffect[Event, State]): PersistentEffect[Event, State] =
      CompositeEffect(effect :: this: Nil)
}

@InternalApi private[akka] final case class CompositeEffect[Event, State](effects: List[PersistentEffect[Event, State]])
    extends PersistentEffect[Event, State] {
    override def andThen(effect: PersistentEffect[Event, State]): PersistentEffect[Event, State] =
      CompositeEffect(effect :: effects)
  }
```